### ISSUE TITLE
TLS: use fewer inotify instances and/or be more resilient to them not working

### ISSUE THREAD

**Issue by nyh** [scylladb/seastar#2513]
Recently some ScyllaDB test runs failed (https://github.com/scylladb/scylladb/issues/21199) with the error:
```
std::system_error (error system:24, could not create inotify instance: Too many open files)
```

It turns out that Seastar's TLS support internally uses inotify to automatically recognize when the certificate files have changed. The error message is misleading - for inotify_init() an EMFILE errno (which gets printed as "Too many open files") does not refer the limit on number of open files, but to the _separate_ per-user limit on the number of inotify instances. This number is configured in `/proc/sys/fs/inotify/max_user_instances` and is often fairly low - e.g., on my Fedora 40 it defaults to just 128. It appears that Seastar creates an inotify instance for each shard, and the ScyllaDB test framework ran many tests in parallel, and the result was running out of inotify instances.

The ScyllaDB testers easily solved this problem by increasing the `/proc/sys/fs/inotify/max_user_instances`, but I'm worried that this problem can hit other Seastar users as well, who won't be aware that Seastar TLS even uses inotify, and that `/proc/sys/fs/inotify/max_user_instances` is so low. I want to propose that we consider two options, perhaps even both:

1. Use fewer inotify instances - perhaps even just once per per process instead of one per shard (e.g., have just shard 0 be responsible for the very rare inotify work). On modern many-shard machines, this will make a big difference.
2. Be forgiving towards failing inotify in TLS - make it a logged warning, not an exception. In that case either the certificate reloading would not work at all - or if we consider it an important feature, we can implement something similar without inotify (e.g., once a minute check if the file changed).

CC @elcallio

**Comment by bhalevy** [comment]
@regevran please assign

**Comment by bhalevy** [comment]
Is it possible the inotify instances are leaking somehow when running the tests via test.py?

**Comment by elcallio** [comment]
No. The problem is that we use a few per shard (one for each file referenced - typically 3-4 per TLS config, say up to 3-4 TLS configs  in scylla.yaml, run 8 shards and then run tests in parallel. This adds up. With ludicrously small default settings we hit the ceiling. 

I've done some work to maybe make the TLS usage in scylla use a shard-0-only solution for TLS inotify, but I got sidetracked. In any case, however you design it, it will be slightly hackish...

**Comment by mykaul** [comment]
https://github.com/scylladb/seastar/commit/6f39b89dbda812c7985918d3d8d6d7f7792b0619 is merged in Seastar, what else do we need?

**Comment by elcallio** [comment]
The issue cannot really be solved on seastar level. https://github.com/scylladb/seastar/commit/6f39b89dbda812c7985918d3d8d6d7f7792b0619 makes it possible however for calling code to avoid shard-multiplication of the inotify usage.

**Comment by elcallio** [comment]
#22425 was merged. It might fix this issue "enough"...

**Comment by nyh** [comment]
> #22425 was merged. It might fix this issue "enough"...

The correct link is https://github.com/scylladb/scylladb/pull/22425

However, this link is to is a patch in Scylla, not in Seastar. I was under the impression (which is why I opened this issue in the Seastar repository) that it was Seastar which is opening these inotify instances, and it was Seastar that does it on every shard.

It seems the solution in https://github.com/scylladb/scylladb/pull/22425 was to only load the certificates on shard 0 because "encryption only does TLS on shard 0 already". Is this fact determined by Seastar or Scylla?
1. If it's just a Scylla implementation that decides to only run TLS on shard 0, then other Seastar users who want to do TLS on all shards (which... sounds reasonable... no?) will still have this problem and it's not solved.
2. If it's Seastar's implementation that only runs TLS on shard 0 (?), then Seastar should know not to open the certificate-reloding inotify things on every shard and should do this only on shard 0, even without patching the caller (Scylla).

**Comment by elcallio** [comment]
The problem cannot really be solved on a seastar level, since the objects in question do not have any "shard-sharing" attributes - they are not sharded<> objects, or even services of any kind. In fact, they have no knowledge of their counterparts/copies across shards. 

#2573 adds some support for _users_ of the objects to possibly reduce inotify footprint (by doing the actual inotify listen only on shard 0 typically), by delegating some event processing to smp callbacks or whatnot, and https://github.com/scylladb/scylladb/pull/22425 uses it on a scylla level.
